p2p/pex/addrbook_test.go

```diff
@@ -36,6 +38,202 @@ func deleteTempFile(fname string) {
+		}
+		addrt := 0
+		if book.IsGood(addr) {
+			addrt = 2
+		} else {
+			addrt = 1
+		}
+
+		if addrt != prevt && prevt != 0 {
+			seqLens = append(seqLens, cnt)
+			seqTypes = append(seqTypes, prevt)
+			cnt = 0
+		}
+		cnt++
+		prevt = addrt
+	}
+	seqLens = append(seqLens, cnt)
+	seqTypes = append(seqTypes, prevt)
+	return seqLens, seqTypes, nil
+}
+
+func TestEmptyAddrBookSelection(t *testing.T) {
+	book := createAddrBookWithMOldAndNNewAddrs(t, 0, 0)
+	addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+	assert.Nil(t, addrs, "expected a nill selection")
+}
+
+func testAddrBookAddressSelection(t *testing.T, b int) {
+	// generate all combinations of old (m) and new (n=b-m) addresses
+	for m := 0; m < b+1; m++ {
+		n := b - m
+		// debug string to show the context for potential failures
+		dbgStr := fmt.Sprintf("book: size %d, number new % d, old %d", b, n, m)
+
+		// create book and get selection
+		book := createAddrBookWithMOldAndNNewAddrs(t, m, n)
+		addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+
+		r := len(addrs)
+		assert.NotNil(t, addrs, "%s - expected a non-nill selection", dbgStr)
+		assert.NotZero(t, r, "%s - expected at least one address in selection", dbgStr)
+
+		// verify that the number of old and new addresses are the ones expected
+		numOld, numNew := countOldAndNewInAddrs(addrs, book)
+		assert.Equal(t, r, numOld+numNew, "%s - expected selection completely filled", dbgStr)
+
+		// Given:
+		// n - num new addrs, m - num old addrs
+		// k - num new addrs expected in the beginning (based on bias %)
+		// i=min(n, k), j=min(m, r-i)
+		//
+		// We expect this layout:
+		// indices:      0...i-1   i...i+j-1    i+j...r
+		// addresses:    N0..Ni-1  O0..Oj-1     Ni...
+		//
+		// There is at least one partition and at most three.
+		k := percentageOfNum(biasToSelectNewPeers, r)
+		i := cmn.MinInt(n, k)
+		j := cmn.MinInt(m, r-i)
+		expOld := j
+		expNew := r - j
+
+		// Verify that the number of old and new addresses are as expected
+		if numNew < expNew || numNew > expNew {
+			t.Fatalf("%s - expected new addrs %d, got %d", dbgStr, expNew, numNew)
+		}
+		if numOld < expOld || numOld > expOld {
+			t.Fatalf("%s - expected old addrs %d, got %d", dbgStr, expOld, numOld)
+		}
+
+		// Verify that the order of addresses is as expected
+		// Get the sequence types and lengths of the selection
+		seqLens, seqTypes, err := analyseSelectionLayout(book, addrs)
+		assert.NoError(t, err, "%s - expected a non-nill selection", dbgStr)
+
+		// Build a list with the expected lengths of partitions and another with the expected types, e.g.:
+		// expseqLens = [10, 22], expseqTypes = [1, 2]
+		// means we expect 10 new (type 1) addresses followed by 22 old (type 2) addresses.
+		var expSeqLens []int
+		var expSeqTypes []int
+		if j == 0 {
+			// all new
+			expSeqLens = []int{r}
+			expSeqTypes = []int{1}
+		} else if i == 0 {
+			// all old
+			expSeqLens = []int{r}
+			expSeqTypes = []int{2}
+		} else if r-i-j == 0 {
```

Comment:
hard to follow :( maybe you can use a switch and add a comment for each case

```go
switch {
  case j == 0: // all new addresses
  case i == 0: // all old addresses
  case r-i-j == 0: // ????
  default: // ????
}
```

Reply:
see https://github.com/tendermint/tendermint/pull/3232#discussion_r253740405
if you get rid of one symbol variables, you might not need `a comment for each case` at all!